kernel/sched.c

```diff
@@ -728,19 +746,29 @@ static void z_thread_halt(struct k_thread *thread, k_spinlock_key_t key,
 
 	if (is_halting(thread) && (thread != _current)) {
 		if (arch_is_in_isr()) {
+			_current_cpu->spinning_on = cpu;
+
 			/* ISRs can only spin waiting another CPU */
 			k_spin_unlock(&sched_spinlock, key);
 			while (is_halting(thread)) {
 			}
```

Comment:
If I am following you correctly, we are no worse off on _that_ race than we were before. For it to occur, this ISR would have to be interrupted by something of even higher priority **and** take long enough for it to occur that the other CPU could context switch in a new thread **and** have some thread re-use the aborted thread's resources. Unlikely, but possible.

I _think_ I see a way to avoid spinning on the `is_halting(thread)`. There may be a simpler method available, but what I have in mind is a k_thread field that contains a bitmap whose set bits indicate which CPUs are spinning in an ISR waiting for the thread to halt and a _cpu field indicating that the target thread is entering a halted (suspended or dead) state. Instead of spinning on is_halting(), the ISR can spin on the new _cpu field until the halting thread indicates it can stop spinning.